### PR TITLE
replication: Try to reinstate entries that have been truncated away

### PR DESCRIPTION
**Pull request by cole-miller**
Experiment with fixing #470

### CODE REVIEW

**Comment by cole-miller** [comment]
please test downstream

**Comment by cole-miller** [comment]
please test downstream

**Comment by cole-miller** [comment]
please test downstream

**Comment by cole-miller** [comment]
Test results look encouraging!

**Comment by cole-miller** [comment]
please test downstream

**Comment by cole-miller** [comment]
please test downstream

**Comment by calvin2021y** [comment]
no update for so long, any  strategic adjustment for this project?

**Comment by freeekanayaka** [comment]
> no update for so long, any strategic adjustment for this project?

I'm sure Mathieu or Cole will be able to speak about that, but FWIW there's now also https://github.com/cowsql/raft, in which I've been fixing a few bugs surfaced by Jepsen tests and that I plan to incrementally move to a v1 version, with a design more or less described by #430 (backward compatibilty will be maintained for the foreseeable future though).

**Comment by cole-miller** [comment]
canonical/raft is still maintained, but the dqlite team has been focused on a broad redesign of dqlite and libraft that aims to address some scalability problems with the current architecture. We're planning to implement that new design during the coming six-month cycle, and we'll publicize our planned changes to dqlite and libraft once the remaining details have been worked out. On the libraft side, our highest priority is making the library work well as a backend for dqlite, although we will maintain the API boundary between the two and avoid introducing any dqlite-specific code. We're not making it a goal to keep libraft backwards-compatible.

Basically, we're happy to see any individuals or projects using libraft, and we plan to keep maintaining it as a separate library, but the direction for it is going to be set by our plans for dqlite, which are going to dictate some big changes that likely won't be convenient for non-dqlite downstreams.

**Comment by calvin2021y** [comment]
[cowsql/raft](https://github.com/cowsql/raft)  look very nice, I will try it. thanks you for this great project and always patient 
communication opportunities. 

For me reliability is more important than compatibility,  since I can always stop the cluster and dump data to upgrade into new version.

High performance comes after reliability, before compatibility.  

Thanks to the canonical /raft team for the hard work.

**Comment by freeekanayaka** [comment]
> [cowsql/raft](https://github.com/cowsql/raft) look very nice, I will try it. thanks you for this great project and always patient communication opportunities.
> 
> For me reliability is more important than compatibility, since I can always stop the cluster and dump data to upgrade into new version.

The on-disk data format will likely not change in v1 of cowsql/raft (or change in minor backward-compatible ways), so no need for users to dump the data or perform data migration. What will change substantially is the C API, which will become more generic and flexible, with better support for disk-based FSMs, asynchronous FSMs, chunked snapshots, custom snapshotting approaches, etc. A compatibility layer will be kept around, so applications using the v0 API will have a very long code upgrade window to migrate from v0 to v1.

> High performance comes after reliability, before compatibility.

Reliability will be a focus too of cowsql/raft, with a requirement to keep all existing upstream and downstream tests (Jepsen, cowsql/dqlite, incus/lxd) always green throughout the development of v1. With the new API in place, there should be more room to improve performance as well, for example by using `io_uring` in very effective and efficient ways. It will take quite some time to get there though.

> Thanks to the canonical /raft team for the hard work.

You're welcome!

**Comment by cole-miller** [comment]
Fixes #470. To sum up: libraft's in-memory log implementation can only track a single log entry for each (index, term) pair. We also need to keep log entries around as long as they're referenced by an `AppendEntries` message that's waiting to be sent. This leads to a situation where a leader steps down suddenly because its disk write for an entry fails, but it's still holding on to that entry in its in-memory log. Another node that received the entry from the original leader can win the next election and send the original leader that same entry, and the original leader will encounter a collision when it tries to append the entry to its log for a second time.

The fix is to have the original leader notice in this situation that the entry it's just received is already tracked by the in-memory log, and just increment the refcount of that existing entry.

File: src/replication.c [hunk 1]
```diff
@@ -1171,6 +1171,21 @@ int replicationAppend(struct raft *r,
      * that we issue below actually completes.  */
     for (j = 0; j < n; j++) {
         struct raft_entry *entry = &args->entries[i + j];
+
```
Comment (MathieuBordere) [inline src/replication.c:1174]: Reasoning out loud here:

The leader has tried to replicate a log entry say with data `D1` at term `T1` and index `I1`, replicates to a majority of followers doesn't hear back from any of them, disk write fails, and now we hit this logic and the leader tries to replicate a new log entry with data `D2` and term `T1` and `I1`, disk write succeeds and leader crashes. I think we now have a situation where there are different log entries for the same term and index.

File: src/log.h [hunk 2]
```diff
@@ -27,6 +27,8 @@ struct raft_entry_ref
     raft_term term;              /* Term of the entry being ref-counted. */
     raft_index index;            /* Index of the entry being ref-counted. */
     unsigned short count;        /* Number of references. */
+    struct raft_buffer buf;      /* Private buffer for the tracked entry */
```
Comment (MathieuBordere) [inline src/log.h:30]: The comment is descriptive, but it would probably be more useful to add the rationale why `buf` and `batch` are added.

**Comment by freeekanayaka** [comment]
FWIW, if it could serve as inspiration I've worked around this issue with this approach https://github.com/cowsql/raft/pull/101. Probably not ideal, but good enough for getting rid of the bug and the assertion failure (I'm planning a broader fix later).

File: src/log.c [hunk 3]
```diff
@@ -480,6 +487,42 @@ static int ensureCapacity(struct raft_log *l)
...
+
+    for (slot = bucket; slot != NULL; slot = slot->next) {
+        if (slot->term == term) {
+            slot->count++;
+            l->back++;
+            l->back %= l->size;
+            entry = &l->entries[l->back];
+            entry->term = term;
+            entry->type = type;
+            entry->buf = slot->buf;
```
Comment (MathieuBordere) [inline src/log.c:517]: are `buf` and `batch` of the previous entry properly cleaned up when doing this?

Comment (cole-miller) [reply src/log.c:517]: They are cleaned up when the old entry is truncated away from the logical log, in removeSuffix. At this point in the code those fields are (logically) uninitialized.

File: src/fixture.c [hunk 4]
```diff
@@ -262,6 +274,11 @@ static void ioFlushAppend(struct io *s, struct append *append)
         goto done;
     }
 
+    if (diskFaultTick(s)) {
```
Comment (MathieuBordere) [inline src/fixture.c:277]: `ioFaultTick` above provides similar functionality. Can you look if it's not too much pain to convert the tests that rely on `ioFaultTick` to simulate a disk failure to `diskFaultTick`?

Comment (cole-miller) [reply src/fixture.c:277]: Looking into this now

Comment (cole-miller) [reply src/fixture.c:277]: Done!

Comment (cole-miller) [reply src/fixture.c:277]: Not really, it would be better to have diskFaultTick and then a separate one to simulate network failures. The unified ioFaultTick is really not great imo and I would be happy to get rid of it.

Comment (cole-miller) [reply src/fixture.c]: This is implemented
